### PR TITLE
Merge CClosure.inductive_subst and Reduction.inductive_subst

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Bench just in case https://gitlab.com/coq/coq/-/jobs/1611898166

**Comment by ppedrot** [comment]
@SkySkimmer can you rebase and rerun the bench? Most of it didn't run it seems.

**Comment by SkySkimmer** [comment]
https://gitlab.com/coq/coq/-/jobs/1615649776

**Comment by SkySkimmer** [comment]
~~~
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-unimath │ 4580.57  4628.94  -1.04 │ 12764330557624  12901708949673  -1.06 │ 24300210287416  24305328247776  -0.02 │ 3695744  3695740   0.00 │   0    0    nan │
│                   coq-verdi │   84.33    85.21  -1.03 │   234828141941    236180658143  -0.57 │   324148897694    324098451134   0.02 │  810260   810236   0.00 │   0    0    nan │
│                  coq-stdlib │  399.66   403.06  -0.84 │  1118713673046   1126472748498  -0.69 │  1419670274757   1419890839181  -0.02 │  593372   593380  -0.00 │   0    0    nan │
│                  coq-geocoq │ 1251.06  1260.70  -0.76 │  3488742273589   3514897982978  -0.74 │  5125389995959   5123862229863   0.03 │  984356   981948   0.25 │   0    0    nan │
│       coq-engine-bench-lite │  283.39   285.20  -0.63 │   796634792287    802475306714  -0.73 │  1455654112785   1458332431208  -0.18 │ 1411028  1411168  -0.01 │   0    0    nan │
│      coq-mathcomp-character │  140.82   141.61  -0.56 │   392248456718    394079196951  -0.46 │   540619217259    540437109619   0.03 │  711480   711648  -0.02 │   0    0    nan │
│               coq-fiat-core │   90.66    91.10  -0.48 │   257833968198    257731618538   0.04 │   327556518009    327458267621   0.03 │  470964   470960   0.00 │   0    0    nan │
│               coq-fourcolor │ 2517.19  2528.79  -0.46 │  7021050158992   7041410740163  -0.29 │ 12181193529033  12180959511861   0.00 │  724184   724384  -0.03 │   0    0    nan │
│            coq-math-classes │  164.38   165.06  -0.41 │   460812716485    462400696404  -0.34 │   602075755380    602105702662  -0.00 │  482012   481828   0.04 │   0    0    nan │
│          coq-mathcomp-field │  210.19   211.04  -0.40 │   585923367350    588917090297  -0.51 │   876702222418    876266201451   0.05 │  655200   655976  -0.12 │   0    0    nan │
│                    coq-hott │  266.60   267.62  -0.38 │   748552088711    751378044949  -0.38 │  1109812109320   1109774812103   0.00 │  585304   585152   0.03 │   0    0    nan │
│       coq-mathcomp-solvable │  161.45   162.02  -0.35 │   450302368989    451364821755  -0.24 │   627794606564    627636838436   0.03 │  689232   690372  -0.17 │   0    0    nan │
│                coq-bedrock2 │  289.69   290.46  -0.27 │   806871858226    809438431195  -0.32 │  1433584546991   1433611399297  -0.00 │  729252   729392  -0.02 │   0    0    nan │
│  coq-performance-tests-lite │ 1377.70  1380.50  -0.20 │  3838061588690   3845222175591  -0.19 │  6226841380936   6227834162813  -0.02 │ 2057300  2057256   0.00 │   0    0    nan │
│        coq-mathcomp-algebra │  131.55   131.81  -0.20 │   366523938513    367531267656  -0.27 │   466970221564    466927002807   0.01 │  553264   553536  -0.05 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1029.02  1031.04  -0.20 │  2864987801202   2874728149275  -0.34 │  4520751879169   4520375818514   0.01 │  928744   896156   3.64 │   0    0    nan │
│               coq-perennial │ 7021.61  7035.00  -0.19 │ 19560285922562  19594813547784  -0.18 │ 29829436618841  29833863550135  -0.01 │ 3138188  3142352  -0.13 │   0    0    nan │
│             coq-lambda-rust │ 1178.96  1180.87  -0.16 │  3283665327970   3289090817239  -0.16 │  4695409445225   4694311231863   0.02 │ 1083244  1082536   0.07 │   0    0    nan │
│            coq-fiat-parsers │  568.34   569.18  -0.15 │  1592488287086   1593658914327  -0.07 │  2429330509726   2428545949121   0.03 │ 3155532  3155436   0.00 │   0    0    nan │
│                   coq-flocq │  132.59   132.78  -0.14 │   368788049758    369994275498  -0.33 │   443130298526    443017272380   0.03 │  930112   927844   0.24 │   0    0    nan │
│              coq-verdi-raft │ 1029.28  1030.67  -0.13 │  2867605956601   2867799756268  -0.01 │  4067587837568   4067115491040   0.01 │ 1230196  1229524   0.05 │   0    0    nan │
│              coq-coquelicot │   59.84    59.92  -0.13 │   165917041635    166353958143  -0.26 │   199055349075    198996800409   0.03 │  684232   684684  -0.07 │   0    0    nan │
│                    coq-core │   99.47    99.60  -0.13 │   276404055133    276469096957  -0.02 │   426390648169    426616874646  -0.05 │  244752   244772  -0.01 │   0    0    nan │
│                   coq-color │  454.89   455.39  -0.11 │  1272624250787   1273480698463  -0.07 │  1579790664635   1579599798185   0.01 │ 1147656  1148064  -0.04 │   0    0    nan │
│                    coq-corn │ 1421.68  1422.44  -0.05 │  3961936352923   3963864951764  -0.05 │  5691682548093   5691327737548   0.01 │  863296   862560   0.09 │   0    0    nan │
│             coq-fiat-crypto │ 4793.14  4794.98  -0.04 │ 13324616960601  13325558574674  -0.01 │ 23521587001871  23493666859089   0.12 │ 2220280  2220012   0.01 │   0    0    nan │
│                coq-coqprime │  320.54   320.60  -0.02 │   894421483346    894367448814   0.01 │  1548300425028   1547159265946   0.07 │  826228   826072   0.02 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │ 1592.08  1590.07   0.13 │  4433165361493   4425245299563   0.18 │  7005740880144   7004327536225   0.02 │ 1136524  1136568  -0.00 │   0    0    nan │
│      coq-mathcomp-ssreflect │   45.16    45.04   0.27 │   126020130813    125776444745   0.19 │   143665884420    143658619540   0.01 │  521476   521148   0.06 │   0    0    nan │
│       coq-mathcomp-fingroup │   42.53    42.39   0.33 │   118165336223    118232169208  -0.06 │   156547983903    156508768415   0.03 │  476036   474908   0.24 │   0    0    nan │
│                 coq-coqutil │   51.83    51.63   0.39 │   145576043428    145229870004   0.24 │   179862366354    179792075887   0.04 │  595208   595536  -0.06 │   0    0    nan │
│                coq-rewriter │  658.81   655.24   0.54 │  1835591772649   1826853390856   0.48 │  2814235087491   2811917316782   0.08 │ 1058760  1058652   0.01 │   0    0    nan │
│                 coq-bignums │   49.69    49.39   0.61 │   138133828022    138096067665   0.03 │   175877065500    175797457207   0.05 │  460440   460240   0.04 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
~~~

File: kernel/reduction.ml [hunk 1]
```diff
@@ -701,9 +686,9 @@ and eqappr cv_pb l2r infos (lft1,st1) (lft2,st2) cuniv =
         let nargs = inductive_cumulativity_arguments ind in
         convert_inductives CONV ind nargs u1 u2 cuniv
       in
-      let pms1 = Array.map_to_list (fun c -> mk_clos e1 c) pms1 in
-      let pms2 = Array.map_to_list (fun c -> mk_clos e2 c) pms2 in
-      let cuniv = List.fold_right2 fold pms1 pms2 cuniv in
+      let pms1 = Array.map (fun c -> mk_clos e1 c) pms1 in
```
Comment (ppedrot) [inline kernel/reduction.ml:689]: Use `CClosure.mk_clos_vect` here and below.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**